lighthouse-core/audits/metrics.js

```diff
@@ -39,66 +39,104 @@ class Metrics extends Audit {
-      });
-    }
+      // Include all timestamps of interest from trace of tab
+      observedNavigationStart: traceOfTab.timings.navigationStart,
+      observedNavigationStartTs: traceOfTab.timestamps.navigationStart,
+      observedFirstPaint: traceOfTab.timings.firstPaint,
+      observedFirstPaintTs: traceOfTab.timestamps.firstPaint,
+      observedFirstContentfulPaint: traceOfTab.timings.firstContentfulPaint,
+      observedFirstContentfulPaintTs: traceOfTab.timestamps.firstContentfulPaint,
+      observedFirstMeaningfulPaint: traceOfTab.timings.firstMeaningfulPaint,
+      observedFirstMeaningfulPaintTs: traceOfTab.timestamps.firstMeaningfulPaint,
+      observedTraceEnd: traceOfTab.timings.traceEnd,
+      observedTraceEndTs: traceOfTab.timestamps.traceEnd,
+      observedLoad: traceOfTab.timings.load,
+      observedLoadTs: traceOfTab.timestamps.load,
+      observedDomContentLoaded: traceOfTab.timings.domContentLoaded,
+      observedDomContentLoadedTs: traceOfTab.timestamps.domContentLoaded,
 
-    // Include all timestamps of interest from trace of tab
-    const timingsEntries = /** @type {Array<[keyof LH.Artifacts.TraceTimes, number]>} */
-      (Object.entries(traceOfTab.timings));
-    for (const [traceEventName, timing] of timingsEntries) {
-      const uppercased = traceEventName.slice(0, 1).toUpperCase() + traceEventName.slice(1);
-      const metricName = `observed${uppercased}`;
-      const timestamp = traceOfTab.timestamps[traceEventName];
-      metrics.push({metricName, timing, timestamp});
-    }
-
-    // Include some visual metrics from speedline
-    metrics.push({
-      metricName: 'observedFirstVisualChange',
-      timing: speedline.first,
-      timestamp: (speedline.first + speedline.beginning) * 1000,
-    });
-    metrics.push({
-      metricName: 'observedLastVisualChange',
-      timing: speedline.complete,
-      timestamp: (speedline.complete + speedline.beginning) * 1000,
-    });
-    metrics.push({
-      metricName: 'observedSpeedIndex',
-      timing: speedline.speedIndex,
-      timestamp: (speedline.speedIndex + speedline.beginning) * 1000,
-    });
+      // Include some visual metrics from speedline
+      observedFirstVisualChange: speedline.first,
+      observedFirstVisualChangeTs: (speedline.first + speedline.beginning) * 1000,
+      observedLastVisualChange: speedline.complete,
+      observedLastVisualChangeTs: (speedline.complete + speedline.beginning) * 1000,
+      observedSpeedIndex: speedline.speedIndex,
+      observedSpeedIndexTs: (speedline.speedIndex + speedline.beginning) * 1000,
+    };
 
-    const headings = [
-      {key: 'metricName', itemType: 'text', text: 'Name'},
-      {key: 'timing', itemType: 'ms', granularity: 10, text: 'Value (ms)'},
-    ];
+    for (const [name, value] of Object.entries(metrics)) {
+      const key = /** @type {keyof UberMetricsItem} */ (name);
+      if (typeof value === 'undefined') {
+        delete metrics[key];
```

Comment:
maybe copy defined values to a new `metrics` object to return rather than delete from the original?

Reply:
any particular reason for this? I'll end up just creating an empty object with no type and casting at the end which seems sad 😢

Reply:
yeah I guess it's not the end of the world to leave them undefined, will do